discovery/manager.go

```diff
@@ -132,43 +128,48 @@ func (m *Manager) startProvider(ctx context.Context, poolKey poolKey, worker Dis
 
 	go worker.Run(ctx, updates)
 	go m.runProvider(ctx, poolKey, updates)
-	go m.runUpdater(ctx)
 }
 
 func (m *Manager) runProvider(ctx context.Context, poolKey poolKey, updates chan []*targetgroup.Group) {
+	ticker := time.NewTicker(5 * time.Second)
+	defer ticker.Stop()
+
+	updateReceived := make(chan struct{}, 1)
+
 	for {
 		select {
 		case <-ctx.Done():
 			return
 		case tgs, ok := <-updates:
-			// Handle the case that a target provider exits and closes the channel
-			// before the context is done.
+			// Handle the case that a target provider(E.g. StaticProvider) exits and
+			// closes the channel before the context is done.
+			// This will prevent sending the updates to the receiver so we send them before exiting.
 			if !ok {
+				select {
+				case m.syncCh <- m.allGroups():
+				default:
+					level.Debug(m.logger).Log("msg", "discovery receiver's channel was full")
+				}
 				return
 			}
 			m.updateGroup(poolKey, tgs)
-			m.recentlyUpdatedMtx.Lock()
-			m.recentlyUpdated = true
-			m.recentlyUpdatedMtx.Unlock()
-		}
-	}
-}
-
-func (m *Manager) runUpdater(ctx context.Context) {
-	ticker := time.NewTicker(5 * time.Second)
-	defer ticker.Stop()
 
-	for {
-		select {
-		case <-ctx.Done():
-			return
-		case <-ticker.C:
-			m.recentlyUpdatedMtx.Lock()
-			if m.recentlyUpdated {
-				m.syncCh <- m.allGroups()
-				m.recentlyUpdated = false
+			// Signal that there was an update.
+			select {
+			case updateReceived <- struct{}{}:
+			default:
+			}
+		case <-ticker.C: // Some discoverers send updates too often so we send these to the receiver once every 5 seconds.
+			select {
```

Comment:
Isn't it possible that the provider loop gets stuck her forever? I'd have expected a default branch to handle the case that there hasn't been any update / a new message on the `updateReceived` channel. Or what am I missing?

Reply:
yes you are right, updated. The default branch now also removes the need for `case <-ctx.Done():` as this will be caught  at the top iteration.